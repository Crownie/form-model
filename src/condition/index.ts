export {Condition} from './condition';
export {ConditionBuilder} from './condition-builder';
export {ConditionGroupBuilder} from './condition-group-builder';
