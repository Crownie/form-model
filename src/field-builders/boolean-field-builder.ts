import {FieldBuilder} from './field-builder';

export class BooleanFieldBuilder extends FieldBuilder {
  constructor(label: string, type: 'switch' | 'checkbox' = 'checkbox') {
    super(type, label, 'boolean');
  }
}
