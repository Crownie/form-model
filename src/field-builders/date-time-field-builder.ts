import {FieldBuilder} from './field-builder';

export class DateTimeFieldBuilder extends FieldBuilder {
  constructor(label: string) {
    super('date-time', label, 'date');
    this.format = 'MM/DD/YYYY hh:mm A';
    this.schemaConfig.nullable = true;
  }

  dateOnly(format: string = 'MM/DD/YYYY'): this {
    this.type = 'date';
    this.format = format;
    return this;
  }

  timeOnly(format: string = 'hh:mm A'): this {
    this.type = 'time';
    this.format = format;
    return this;
  }
}
