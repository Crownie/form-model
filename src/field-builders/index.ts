export * from './boolean-field-builder';
export * from './date-time-field-builder';
export * from './field-builder';
export * from './list-field-builder';
export * from './money-field-builder';
export * from './object-field-builder';
export * from './select-field-builder';
export * from './slider-field-builder';
export * from './text-field-builder';
