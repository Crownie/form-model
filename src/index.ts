export * from './condition';
export * from './field-builders';
export * from './field';
export * from './form';
export * from './wiz-form';
