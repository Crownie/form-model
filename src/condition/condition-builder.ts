import {ConditionGroupBuilder} from './condition-group-builder';
import {LogicalOperator, Operator} from './condition';

export class ConditionBuilder {
  public operator?: Operator;
  public left?: any;
  public right?: any;
  public conditions?: ConditionBuilder[];
  private readonly conditionGroup: ConditionGroupBuilder | null = null;

  constructor(conditionGroup: ConditionGroupBuilder | null, left: any) {
    this.conditionGroup = conditionGroup;
    this.left = left;
  }

  equalTo(val): ConditionGroupBuilder {
    this.operator = '$eq';
    this.right = val;
    return this.conditionGroup!;
  }

  greaterThan(val): ConditionGroupBuilder {
    this.operator = '$gt';
    this.right = val;
    return this.conditionGroup!;
  }

  lessThan(val): ConditionGroupBuilder {
    this.operator = '$lt';
    this.right = val;
    return this.conditionGroup!;
  }

  greaterThanOrEqualTo(val): ConditionGroupBuilder {
    this.operator = '$gte';
    this.right = val;
    return this.conditionGroup!;
  }

  lessThanOrEqualTo(val): ConditionGroupBuilder {
    this.operator = '$lte';
    this.right = val;
    return this.conditionGroup!;
  }

  isIn(vals: (string | number)[]): ConditionGroupBuilder {
    this.operator = '$in';
    this.right = vals;
    return this.conditionGroup!;
  }

  logic(operator: LogicalOperator, conditions) {
    this.operator = operator;
    this.conditions = conditions;
    this.left = undefined;
    this.right = undefined;
  }
}
