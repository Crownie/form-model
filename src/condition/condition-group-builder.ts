import {Condition, LogicalOperator} from './condition';
import {ConditionBuilder} from './condition-builder';

export class ConditionGroupBuilder {
  private readonly conditionBuilder: ConditionBuilder;

  constructor(private logic: LogicalOperator) {
    this.conditionBuilder = new ConditionBuilder(this, null);
  }

  static all(): ConditionGroupBuilder {
    return new ConditionGroupBuilder('$and');
  }

  static any(): ConditionGroupBuilder {
    return new ConditionGroupBuilder('$or');
  }

  when(val): ConditionBuilder {
    const condition = new ConditionBuilder(this, val);
    this.conditionBuilder.logic(this.logic, [condition]);
    return condition;
  }

  nest(group: ConditionGroupBuilder) {
    this.conditionBuilder.conditions!.push(group.conditionBuilder);
    return this;
  }

  build(): Condition {
    return new Condition(this.conditionBuilder);
  }
}
