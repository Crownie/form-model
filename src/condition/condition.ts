import lodashGet from 'lodash.get';
import {ConditionBuilder} from './condition-builder';

export class Condition implements ConditionAttributes {
  public operator?: Operator;
  public left?: any;
  public right?: any;
  public conditions?: Condition[];

  constructor(builder: ConditionBuilder | Condition | ConditionAttributes) {
    this.serialize(builder);
  }

  static processCondition(data, condition: Condition): boolean {
    const {conditions = []} = condition;
    switch (condition.operator) {
      case '$and':
        return conditions.every((item) => this.processCondition(data, item));
      case '$or':
        return conditions.some((item) => this.processCondition(data, item));
      case '$eq':
        return lodashGet(data, condition.left) === condition.right;
      case '$gt':
        return lodashGet(data, condition.left) > condition.right;
      case '$lt':
        return lodashGet(data, condition.left) < condition.right;
      case '$gte':
        return lodashGet(data, condition.left) >= condition.right;
      case '$lte':
        return lodashGet(data, condition.left) <= condition.right;
      case '$in':
        const arr = Array.isArray(condition.right) ? condition.right : [];
        return arr.includes(lodashGet(data, condition.left));
    }
    return false;
  }

  evaluate(data: {[key: string]: any}): boolean {
    return Condition.processCondition(data, this);
  }

  serialize(obj: ConditionBuilder | any) {
    this.operator = obj.operator;
    this.left = obj.left;
    this.right = obj.right;
    this.conditions = obj.conditions
      ? obj.conditions.map((condition) => {
          return new Condition(condition);
        })
      : obj.conditions;
  }
}

export type LogicalOperator = '$and' | '$or';
export type Operator =
  | '$eq'
  | '$gt'
  | '$lt'
  | '$gte'
  | '$lte'
  | '$in'
  | LogicalOperator;

export interface ConditionAttributes {
  operator?: Operator;
  left?: any;
  right?: any;
  conditions?: Condition[];
}
