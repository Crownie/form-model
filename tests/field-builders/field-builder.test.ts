import {FieldBuilder} from '../../src/field-builders';
import {Condition} from '../../src/condition';

it('FieldBuilder', () => {
  const field = new FieldBuilder('ts', '', 'string')
    .setWidth({xs: 12, sm: 2}, true)
    .required()
    .min(1)
    .max(2)
    .equalToField('other-field', 'must match other field')
    .email()
    .hideWhen(
      new Condition({operator: '$eq', left: 'some-field', right: 'some-value'}),
    )
    .disableWhen(
      new Condition({operator: '$eq', left: 'some-field', right: 'some-value'}),
    )
    .matches(/.*/, 'must match pattern')
    .build();

  expect(field.min).toEqual(field.getSchemaConfig().min!.val);
  expect(field.max).toEqual(field.getSchemaConfig().max!.val);
  expect(field).toMatchSnapshot();
});
