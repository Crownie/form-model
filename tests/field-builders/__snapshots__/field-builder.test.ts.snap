// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldBuilder 1`] = `
Field {
  "conditionalDisable": Condition {
    "conditions": undefined,
    "left": "some-field",
    "operator": "$eq",
    "right": "some-value",
  },
  "conditionalHide": Condition {
    "conditions": undefined,
    "left": "some-field",
    "operator": "$eq",
    "right": "some-value",
  },
  "label": "",
  "meta": Object {},
  "noWrap": true,
  "schemaConfig": Object {
    "defaultVal": "",
    "email": Object {
      "errorMessage": "Not a valid email",
      "val": "Not a valid email",
    },
    "equalToField": Object {
      "errorMessage": "must match other field",
      "val": "other-field",
    },
    "matches": Object {
      "errorMessage": "must match pattern",
      "val": /\\.\\*/,
    },
    "max": Object {
      "errorMessage": "Too long",
      "val": 2,
    },
    "min": Object {
      "errorMessage": "Too short",
      "val": 1,
    },
    "required": Object {
      "errorMessage": "Required",
      "val": "Required",
    },
    "type": "string",
  },
  "type": "ts",
  "width": Object {
    "sm": 2,
    "xs": 12,
  },
}
`;
